### PR TITLE
[namedtensor] fix autograd support for torch.mean(tensor, dimname)

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29199 [namedtensor] fix autograd support for torch.mean(tensor, dimname)**

Previously, we called `native::mean_cpu_gpu` inside `mean(Tensor, Dimname)`;
`native::mean_cpu_gpu` is not supported by autograd. This PR replaces
`native::mean_cpu_gpu` with `at::mean(Tensor, int)` so that the dimname
overload can piggyback off of autograd support for `at::mean(Tensor,
int)`.

Also added tests (those didn't exist before) for autograd support for
named tensor reduction functions.

Test Plan:
- `python test/test_namedtensor.py -v`

Differential Revision: [D18334617](https://our.internmc.facebook.com/intern/diff/D18334617)

### CODE REVIEW

File: test/test_namedtensor.py [hunk 1]
```diff
@@ -1054,6 +1062,11 @@ def test_simple_reduce(op, device):
             with self.assertRaisesRegex(RuntimeError, 'Name \'H\' not found'):
                 op(t, 'H')
 
+        def test_autograd_supports_dimname_overload(op, device):
+            t = torch.empty(2, 3, 5, names=('N', 'C', 'L'), device=device, requires_grad=True)
+            sum_all_outputs(op(t, 'C')).backward()
+            self.assertTrue(t.grad is not None)
```
Comment (nairbv) [inline test/test_namedtensor.py]: can we do anything to validate the value of the gradient? would it be helpful to validate the grad against the equivalent op on an unnamed tensor?

Comment (nairbv) [reply test/test_namedtensor.py]: nit: btw I think there's a self.assertIsNotNone()

Comment (zou3519) [reply test/test_namedtensor.py]: > can we do anything to validate the value of the gradient?

We can, but see the next point

> would it be helpful to validate the grad against the equivalent op on an unnamed tensor?

Because autograd ignores all names on tensors, we don't need to validate the grad as long as we test that "autograd ignores all names on tensors"

> nit: btw I think there's a self.assertIsNotNone()

Sure I'll use that

**Review by nairbv — APPROVED** [review]

File: test/test_namedtensor.py (hunk 1 above)
Comment (nairbv) [reply test/test_namedtensor.py]: >autograd ignores all names on tensors 

will that ever not be the case for any kinds of gradients?

I think you had also mentioned an issue with the correctness of the mean gradient that we were also fixing, though maybe that's (going to be) a separate PR?

Comment (zou3519) [reply test/test_namedtensor.py]: > > autograd ignores all names on tensors
> 
> will that ever not be the case for any kinds of gradients?

In the future we may design and implement name-aware autograd. This is not going to happen soon. When that happens, we can revisit the testing plan for this.

> I think you had also mentioned an issue with the correctness of the mean gradient that we were also fixing, though maybe that's (going to be) a separate PR?

yeah, that's orthogonal to this PR. It is this one: https://github.com/pytorch/pytorch/issues/29137#issuecomment-549938081

Comment (zou3519) [reply test/test_namedtensor.py]: This PR fixes https://github.com/pytorch/pytorch/issues/28286

Comment (nairbv) [reply test/test_namedtensor.py]: thanks for the clarification
